### DISCUSSION TITLE
Replacing Jetstream teams roles and permissions with Spatie permission team feauture

### DISCUSSION THREAD

**Discussion by kaspernowak** [spatie/laravel-permission#2598] [Q&A]
Hello All,
I originally implemented the teams feature from laravel jetstream into my laravel breeze app, including the role and permissions approach used by jetstream. Now I need to implement global roles and permissions, where I found Spatie Permissions useful. Then I found out that Spatie Permissions can be installed with a teams feature as well, which made me pivot towards using Spatie Permissions for my Team roles and permissions as well.

Now it seems like I have hit a wall, where I am not sure that the Spatie Permissions teams feature can replace the Jetstream teams feature, in the way that Jetstream sets the team roles and permissions in a serviceprovider, where roles can be added or removed and permissions can be changed dynamically, affecting all existing and future teams. 
Where in what seems to be the case for Spatie Permission teams, roles should be created for each team, even though all teams need the same roles and permissions, making it difficult to manage team roles from, say,  something like one centralized service providers boot method.

I am not sure I understand the Spatie Permission teams feauture correctly, but if I understand it correctly, I don't see the point of why anyone would want to use it?

Maybe it is best to keep using the implemented role and permission setup from jetstream for my teams, and use spatie for global roles and permissions?

Does anyone have experience with using both jetstream teams and spatie permission teams, and know if I could successfully pivot to spatie permission with teams, or if I should stick with my existing implementation and live with having to manage roles from 2 different methods in my app?

**Comment by osama300** [comment]
did you find solution?

**Reply by kaspernowak** [reply]
@osama300 I used the jetstream teams roles for my teams, and spatie for global roles and permissions. It was too much work at the time to find one solution that would work for all, and using both works fine in our case.

**Reply by osama300** [reply]
I had a similar situation and managed to combine Jetstream Teams with Spatie Permissions (with teams support).

### My approach:

- I use **global roles** without `team_id`, such as `team_owner`, which apply across the app.
- Each **team can create their own roles** with a `team_id`, such as `Manager`, `Accountant`, etc.
- I ignore Jetstream’s default role management and rely only on Spatie’s roles/permissions.

Global roles
id | team_id | name
----------------------
39 | NULL     | team_owner

// Team-specific role
42 | 1        | (manager)


### Integration with Jetstream Invitations:

Ignore Jetstream’s default role system
Instead of using Jetstream's built-in role string on team invitations,
 I added a new field to the team_invitations table: spatie_role_id
 
id | team_id | email | role | spatie_role_id | created_at | updated_at

When a user accepts a team invitation, I attach the correct Spatie role via a listener:

```php
public function handle(TeamMemberAdded $event): void
{
    $invitation = TeamInvitation::where('team_id', $event->team->id)
                                ->where('email', $event->user->email)
                                ->first();

    if ($invitation && $invitation->spatie_role_id) {
        setPermissionsTeamId($event->team->id);
        $event->user->assignRole($invitation->spatie_role_id);
    }
}


This setup allows me to:

Avoid duplicating the same roles across all teams.

Maintain global roles easily.

Still let each team manage their own set of roles if needed.

**Reply by thilipkumart** [reply]
@osama300 , Hi if you implemented this in your project, could you please share us your GIT repo
